### ISSUE TITLE
Question about "Poison" in today's lecture

### ISSUE THREAD

**Issue by alapha23** [snu-sf-class/swpp202201#15]
I'm not sure I understand `poison` perfectly and I am reading [this slide by Lee on llvm.org](https://llvm.org/devmtg/2020-09/slides/Lee-UndefPoison.pdf) to try to understand it.

According to the slides:
* Problem: Needed a value that represents signed overflow in LLVM IR
* Compiler assumes the source has no UB.
* `poison` is a special value that represents a violation of an assumption
* Each operation either propagates poison or raise UB

My questions are
* Why do we need to represent signed overflow in LLVM IR? Aka. why do we need poison at all? If there's an overflow, it's the programmer's fault. Because as suggested the compiler can assume the source has no UB
* We already have `undef`, could you elaborate a bit on the difference between `undef` and `poison`? Can't we use `undef` to represent all UBs?
* In the example of our lecture today, the UB was caused by the compiler optimization (I think it was code invariant motion basically moving things out of if-else) -- why can't we fix this compiler-caused UB by specifically check the condition when it happens? I'm having some wild thoughts of a few possible options 
     * a. If overflowing is possible to happen due to code invariant motion optimization, then don't do code invariant motion optimization
     * b. If UB can happen and it's not the compiler's fault, then don't do anything. Let the programmers eat what they've planted
     * c. Implement the semantics checking after doing optimizations, then send a warning if there's UB. This helps because we don't have to fix the UB in the compiler, we just tell the programmer there could be one due to the optimization. 

Basically I don't understand the point of `poison` being a necessity. 
What do you think?
Thank you!

**Comment by Sung-HwanLee** [comment]
Hi! I address your questions in the following.
You can find the paper on undef/poison from our lab page: [paper](https://sf.snu.ac.kr/publications/undefllvm.pdf).

### Why do we need signed overflows in LLVM? Why don't we consider them to be UB?
As you pointed out, compilers intend to allow code motion - consider the following example.
```
foo(int a, int n) {
  for (int i = 0; i < n; i++) {
    int x = a + 3;
    ...    // use x
  }
}
```
Here, `int x = a + 3` inside the loop is a "loop invariant code" that repeats the same operation every iteration.
Therefore, compilers want to pull out the invariant from the loop by performing the code motion, resulting in the following code:
```
foo(int a, int n) {
  int x = a + 3;
  for (int i = 0; i < n; i++) {
    ...    // use x
  }
}
```
This transformation is called loop invariant code motion (LICM).
However, LICM is unsound if a signed overflow is UB because it may introduce a signed overflow when `n = 0`.
In particular, calling `foo(INTMAX, 0)` does not raise UB before the transformation, while it does raise UB after because of the overflow.
Therefore, to allow LICM (and some other transformations), instead of regarding overflows as UB, compilers consider an overflow resulting in a special value.

### What is the difference between undef and poison?
Roughly speaking, undef is a set of values, while poison is a special value that propagates itself on arithmetics and raises UB on branching.
For example:
```
poison + 1 = poison
poison < INTMAX = poison
if (poison) ... = UB
```
As illustrated in the slides you mentioned, `undef <= INTMAX` is always true because any integer is less than or equal to `INTMAX`, while `poison <= INTMAX` is `poison`.
There were certain compiler transformations (that LLVM performed) that are unsound under the undef semantics.

### Regarding your suggestions
I am not quite sure what do you mean by b and c.
Let me know if I misunderstood your questions.
- a. Yes, prohibiting certain compiler transformations including LICM is another direction. However, this is performance-critical, for example, when calling the function `foo` with large `n`. Therefore, such a choice is not acceptable in general.
- b. Compilers do not try to do anything about user faults. They just perform optimizations that are sound under the IR semantics. The question is what should be the correct semantics is.
- c. I don't see what do you mean by "implementing the semantics". Compilers should never introduce UB that does not exist in the source program by the definition of "sound transformations". Practically, what can I do if a compiler tells me that "your program is good, there is no UB, but maybe I inserted UB, so this executable may behave arbitrarily"?
